### PR TITLE
Roller Bed-der

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

Refactors some of the procs and logic for roller beds to instead store the bed inside of the structure instead of constantly del'ing and spawning new ones.

## Why It's Good For The Game

Stops a bunch of rather pointless del'ing and spawning of objects.
Also this makes it so family heirloom components aren't yeeted into the void; so I guess you can call this a paramedic power-creep.

## Changelog
:cl:
code: Roller Beds are now more intelligent in how they spawn/delete
/:cl:

### CODE REVIEW

File: code/game/objects/structures/beds_chairs/bed.dm [hunk 1]
```diff
@@ -52,29 +52,41 @@
...
-		var/obj/item/roller/robo/R = W
-		if(R.loaded)
-			to_chat(user, span_warning("You already have a roller bed docked!"))
-			return
+/obj/structure/bed/roller/Initialize(mapload, obj/item/roller/roller)
+	. = ..()
+	if(!roller)
+		roller = new /obj/item/roller(src)
+	src.roller = roller
+	roller.forceMove(src)
```
Comment (Kylerace) [inline code/game/objects/structures/beds_chairs/bed.dm:63]: roller already exists inside of src at this point

File: code/game/objects/structures/beds_chairs/bed.dm [hunk 2]
```diff
@@ -52,29 +52,41 @@
...
+	name = roller.name
+
+/obj/structure/bed/roller/Destroy()
+	if(roller)
+		qdel(roller)
+		roller = null
```
Comment (Kylerace) [inline code/game/objects/structures/beds_chairs/bed.dm:69]:
```suggestion
		QDEL_NULL(roller)
```

File: code/game/objects/structures/beds_chairs/bed.dm [hunk 3]
```diff
@@ -52,29 +52,41 @@
...
-				user.visible_message(span_notice("[user] unbuckles all creatures from [src]."))
-			else
-				user_unbuckle_mob(buckled_mobs[1],user)
-		else
-			R.loaded = src
-			forceMove(R)
-			user.visible_message(span_notice("[user] collects [src]."), span_notice("You collect [src]."))
-		return 1
-	else
+/obj/structure/bed/roller/attackby(obj/item/W, mob/user, params)
```
Comment (Kylerace) [inline code/game/objects/structures/beds_chairs/bed.dm:72]: use a more descriptive name than W please

File: code/game/objects/structures/beds_chairs/bed.dm [hunk 4]
```diff
@@ -52,29 +52,41 @@
 	icon_state = "down"
 	anchored = FALSE
 	resistance_flags = NONE
-	var/foldabletype = /obj/item/roller
+	/// Internal storage for the roller item that created us
+	var/obj/item/roller
```
Comment (Kylerace) [inline code/game/objects/structures/beds_chairs/bed.dm:56]: you dont register for this being qdeleted

**Review by Kylerace — CHANGES_REQUESTED** [review]
im confused on why this is needed. creating a new roller object based on player input is not a performance issue in the slightest, and containing it as a physical object means it then has to be managed. can you elaborate on what this fixes?

**Comment by Kylerace** [comment]
also changelog is for player facing changes

File: code/game/objects/structures/beds_chairs/bed.dm (hunk 3 above)
Comment (ZephyrTFA) [reply code/game/objects/structures/beds_chairs/bed.dm:71]: That's the base definition of the proc.

File: code/game/objects/structures/beds_chairs/bed.dm (hunk 1 above)
Comment (ZephyrTFA) [reply code/game/objects/structures/beds_chairs/bed.dm:63]: Not if it didn't make a new one

File: code/game/objects/structures/beds_chairs/bed.dm (hunk 4 above)
Comment (ZephyrTFA) [reply code/game/objects/structures/beds_chairs/bed.dm:56]: Please suggest so I can web commit, at work tyly

**Comment by Cyberboss** [comment]
I'm with @Kylerace on this one. Creating/deleting an object in response to occasional player input is very CPU light. Keeping the bed around just causes a memory footprint, albeit a small one, and extra code maintenance pressure. I'm not sold on the benefits of this.

**Comment by ZephyrTFA** [comment]
> 
> 
> I'm with @Kylerace on this one. Creating/deleting an object in response to occasional player input is very CPU light. Keeping the bed around just causes a memory footprint, albeit a small one, and extra code maintenance pressure. I'm not sold on the benefits of this.

This is mainly to fix the FamilyHeirloom component being yeeted into the void for Paramedics

**Comment by Kylerace** [comment]
@ZephyrTFA then note it as a fix in the changelog for the bug you intend to fix, players dont need to know about internal only changes to code unless it fixes a bug

**Review by Kylerace — CHANGES_REQUESTED** [review]

**Pull request closed without merging**